---
title: "HapRFN workflow"
author: "Bencze Balazs"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{HapRFN}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### HapRFN Package

The HapRFN package is used to find segments that are Identity by Descent (IBD) in a dataset describing single nucleotide variations (SNVs) of a sample of population (individuals or haplotypes).
The package is split up into multiple dependent functions that have to be used in a specific order. This document describes and explains that order.

## Analysis pipeline

# Preprocessing with vcf2sparse

The function converts a .vcf file to a sparse matrix format. The vcf file can be in text or gzipped format. The gzipped vcf file can be accompanied by an index file, which must have the same name as the vcf file (e.g. chr1.vcf.gz and chr1.vcf.gz.tbi). This function also splits up the dataset into intervals, to reduce the search space for IBD segments (they are usually grouped together).

```r
library(hapRFN)

# The name of the VCF file without the extension (.vcf and .vcf.gz accepted)
fileName <- "name_of_the_vcf_file"

# This is the path to the vcf file
path <- "/the/path/to/the/vcf/file"

# The name of the output file (defaults to the input filename,
# but appends a different extension and it does not overwrite it)
output <- "the_name_of_the_output_file"

# The path to the output of vcf2sparse
outputPath <- "/the/path/to/the/output/file"

# These are the default values, that are used throughout the package
intervalSize <- 10000
shiftSize <- 5000

vcf2sparse(fileName, path, intervalSize, shiftSize, outputFile = output,
           outputPrefixPath = outputPath)
```

Calling this command will split the matrix in multiple intervals, in a sparse matrix format, and generate an annotation, info and individuals file. These files will be used in the subsequent calls.

# Running the HapRFN method

This function iterates over the splitted matrices and applies the HapRFN algorithm to each split. After running RFN it also extracts the IBD segments.

```r
# Set the id number of the GPU
gpuId <- 0

# The number of IBDsegments (biclusters) to detect
biclusters <- 50

annotationFile <- paste0(outputPath, "/", output, "_annot.txt")

# Path to the analysis files
analysisResults <- "/the/path/to/the/result/files"

iterateIntervals(annotationFile = annotationFile, fileName = output,
  shiftSize = shiftSize, intervalSize = intervalSize,
  prefixPath = path, gpuId = gpuId, p = biclusters)

identifyDuplicates(fileName = fileName, shiftSize = shiftSize,
  intervalSize = intervalSize)

anaRes <- analyzeIBDsegments(fileName = fileName, shiftSize = shiftSize,
  intervalSize = intervalSize)
```

# Analysis of the results

```r
print("Number IBD segments:")
print(anaRes$noIBDsegments)
print("Statistics on IBD segment lengths in SNVs (all SNVs in the
IBD segment):")
print(anaRes$avIBDsegmentLengthSNVS)
print("Statistics on IBD segment lengths in bp:")
print(anaRes$avIBDsegmentLengthS)
print("Statistics on number of individuals that share an IBD segment:")
print(anaRes$avnoIndividS)
print("Statistics on number of IBD segment tagSNVs:")
print(anaRes$avnoTagSNVsS)
print("Statistics on MAF of IBD segment tagSNVs:")
print(anaRes$avnoFreqS)
print("Statistics on MAF within the group of IBD segment tagSNVs:")
print(anaRes$avnoGroupFreqS)
print("Statistics on number of changes between major and minor allele frequency:")
print(anaRes$avnotagSNVChangeS)
print("Statistics on tagSNVs per individual that shares an IBD segment:")
print(anaRes$avnotagSNVsPerIndividualS)
print("Statistics on number of individuals that have the minor allele of tagSNVs:")
print(anaRes$avnoindividualPerTagSNVS)

# Load result for interval 50
posAll <- 50 # (50 - 1) * 5000 = 245000: segment 245000 to 255000
start <- (posAll - 1) * shiftSize
end <- start + intervalSize
pRange <- paste0("_", format(start, scientific = FALSE), "_",
  format(end, scientific = FALSE))
load(file = file.path(analysisResults, paste0(output, pRange, "_resAnno", ".Rda")))
IBDsegmentList <- resHapRFN$mergedIBDsegmentList

summary(IBDsegmentList)

# Set this to true, to plot to a file
plotToFile <- T

if (plotToFile) {
    pdf("IBD_segments_interval_50.pdf")
}

# Plot IBD segments in interval 50

matFile <- file.path(outputPath, paste0(fileName, pRange, "_matG.txt"))
plotIBDSegmentList(IBDsegmentList, fileName = matFile, interactive = !plotToFile)

# Plot the first IBD segment in interval 50

IBDsegment <- IBDsegmentList[[1]]

if (plotToFile) {
    pdf("IBD_segment_1_interval_50")
}
plotIBDSegment(IBDsegment, fileName = matFile, interactive = !plotToFile)

if (plotToFile) {
    dev.off()
}

```

